File: DittoDatto-old/packages/shared-types/src/company.ts
import { z } from "zod";
import { IdSchema, DateTimeSchema } from "./common";

export const CompanySchema = z.object({
  id: IdSchema,
  ownerId: IdSchema,
  ownerEmail: z.email().optional(),
  name: z.string().min(1).max(255),
  description: z.string().optional(),
  website: z.url().optional(),
  address: z.string().optional(),
  city: z.string().optional(),
  zip: z.string().optional(),
  country: z.string().min(2).max(255).default("NO"), // ISO country code

  email: z.email().optional(),
  phone: z.string().optional(),
  logoUrl: z.url().optional(),
  tier: z.enum(["free", "premium"]).default("free"),
  slinks: z
    .object({
      fb: z.url().optional(),
      ig: z.url().optional(),
      x: z.url().optional(),
    })
    .optional(),

  // Onboarding
  onboardingStatus: z.enum([
    "not_started",
    "ai_suggested",
    "verified",
    "complete",
  ]),
  aiSuggestedData: z
    .object({
      name: z.string().optional(),
      website: z.url().optional(),
      description: z.string().optional(),
      address: z.string().optional(),
      phone: z.string().optional(),
      suggestedAt: DateTimeSchema,
    })
    .optional(),

  // Feature Flags (Admin-controlled)
  enabledFeatures: z
    .object({
      tableReservation: z.boolean().default(false),
      aiAssistance: z.boolean().default(false),
      ticketSystem: z.boolean().default(false),
      eventSystem: z.boolean().default(false),
    })
    .default({
      tableReservation: false,
      aiAssistance: false,
      ticketSystem: false,
      eventSystem: false,
    }),

  // Store Policy (Admin-controlled)
  storePolicy: z
    .object({
      maxStores: z.number().int().min(0).default(1),
      canCreateOwnStores: z.boolean().default(false),
    })
    .default({
      maxStores: 1,
      canCreateOwnStores: false,
    }),

  // Media Configuration
  mediaConfig: z
    .object({
      defaultTags: z
        .array(z.string())
        .default(["logo", "cover", "staff", "store", "menu", "misc"]),
    })
    .default({
      defaultTags: ["logo", "cover", "staff", "store", "menu", "misc"],
    }),

  // Metadata
  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema,
  managerIds: z.array(IdSchema).optional(),
  memberIds: z.array(IdSchema).optional(),
});
export type Company = z.infer<typeof CompanySchema>;

File: packages/mercury_client/lib/src/models/company.dart
import 'package:json_annotation/json_annotation.dart';

import 'enums.dart';

part 'company.g.dart';

/// Social media and website links for a company.
@JsonSerializable()
class CompanySocialLinks {
  const CompanySocialLinks({this.website, this.fb, this.ig, this.x});

  factory CompanySocialLinks.fromJson(Map<String, dynamic> json) =>
      _$CompanySocialLinksFromJson(json);

  final String? website;
  final String? fb;
  final String? ig;
  final String? x;

  Map<String, dynamic> toJson() => _$CompanySocialLinksToJson(this);
}

/// Store policy config for a company.
@JsonSerializable()
class StorePolicy {
  const StorePolicy({
    this.maxStores = 1,
    this.canCreateOwnStores = false,
  });

  factory StorePolicy.fromJson(Map<String, dynamic> json) =>
      _$StorePolicyFromJson(json);

  @JsonKey(name: 'max_stores')
  final int maxStores;

  @JsonKey(name: 'can_create_own_stores')
  final bool canCreateOwnStores;

  Map<String, dynamic> toJson() => _$StorePolicyToJson(this);
}

/// Feature flags controlling what modules a company has access to.
@JsonSerializable()
class EnabledFeatures {
  const EnabledFeatures({
    this.tableReservation = false,
    this.aiAssistance = false,
    this.ticketSystem = false,
    this.eventSystem = false,
  });

  factory EnabledFeatures.fromJson(Map<String, dynamic> json) =>
      _$EnabledFeaturesFromJson(json);

  @JsonKey(name: 'table_reservation')
  final bool tableReservation;

  @JsonKey(name: 'ai_assistance')
  final bool aiAssistance;

  @JsonKey(name: 'ticket_system')
  final bool ticketSystem;

  @JsonKey(name: 'event_system')
  final bool eventSystem;

  Map<String, dynamic> toJson() => _$EnabledFeaturesToJson(this);
}

/// A company (business) registered on the DittoDatto platform.
@JsonSerializable()
class Company {
  const Company({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    this.email,
    this.phone,
    this.address,
    this.city,
    this.postalCode,
    this.country = 'NO',
    required this.tier,
    required this.onboardingStatus,
    this.ownerEmail,
    required this.ownerId,
    required this.dbSlug,
    this.socialLinks = const CompanySocialLinks(),
    this.storePolicy = const StorePolicy(),
    this.enabledFeatures = const EnabledFeatures(),
    required this.createdAt,
    required this.updatedAt,
  });

  factory Company.fromJson(Map<String, dynamic> json) =>
      _$CompanyFromJson(json);

  final String id;
  final String name;
  final String slug;
  final String? description;
  final String? email;
  final String? phone;
  final String? address;
  final String? city;

  @JsonKey(name: 'zip')
  final String? postalCode;

  final String country;
  final CompanyTier tier;

  @JsonKey(name: 'onboarding_status')
  final OnboardingStatus onboardingStatus;

  @JsonKey(name: 'owner_email')
  final String? ownerEmail;

  @JsonKey(name: 'owner_id')
  final String ownerId;

  @JsonKey(name: 'db_slug')
  final String dbSlug;

  @JsonKey(name: 'social_links')
  final CompanySocialLinks socialLinks;

  @JsonKey(name: 'store_policy')
  final StorePolicy storePolicy;

  @JsonKey(name: 'enabled_features')
  final EnabledFeatures enabledFeatures;

  @JsonKey(name: 'created_at')
  final DateTime createdAt;

  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  Map<String, dynamic> toJson() => _$CompanyToJson(this);
}
